File: spex/viewer/src/show/choreography.ts
/** M64 — runtime choreography. Gap B, closed.
 *
 * `spex brick-assembly` bakes nine bricks settling out of a scattered start
 * into a sequence of point-cloud frames. A1-S04 is that shot, and from here it
 * is *evaluated*, not baked: the same constants, the same seeding, the same
 * curve, computed per frame through `InstanceWriter`.
 *
 * The baked path is not retired. It stays exactly as it is, and it is still
 * the right tool for a quick demo of a scene nobody is going to seek around
 * in. What changes is that the *show* stops using it.
 *
 * # The generator is declared, not expanded
 *
 * The spec asked for the scatter to be turned into a transform track at
 * `show-build` time. That works for nine bricks and not for an Atlas site: a
 * transform track carries one value per keyframe shared across its whole
 * target, so per-instance scatter needs one track *per instance* — five
 * thousand tracks of two keyframes each, several megabytes of JSON, to
 * express two constants and a seed. So the document declares the generator
 * (A1-S04's `seed` cue carries `floatHeightLdu`, `scatterRadiusLdu`, the
 * stagger mode and the easing) and this evaluates it. The resolved document
 * stays the size of a screenplay.
 *
 * # One PRNG, two languages
 *
 * `crates/spex-show/src/choreography.rs` is the other half of this file, and
 * the two are pinned to `docs/fugen/fixtures/assembly-scatter.json` rather
 * than to each other — a fixture they can both fail against, instead of a
 * comparison they could both drift through.
 *
 * splitmix64 rather than `rand::StdRng`, because `StdRng` is ChaCha12 and
 * `rand` makes no promise its stream is stable across versions. See the Rust
 * module's header.
 *
 * # Coordinates
 *
 * The generator works in **LDraw units, Y-down**, because that is the frame
 * the constants were measured in. `startOffsetMm` converts at the same
 * boundary `spex_mesh::bundle::to_output_position` does, and negating Y is
 * the same mirror it is everywhere else in this project.
 */

import { cubicInOut, easingByName, type EasingFn } from './easing';

export const FLOAT_HEIGHT_LDU = 420;
export const SCATTER_RADIUS_LDU = 260;
/** LDraw's own unit. One LDU is 0.4 mm. */
export const LDU_MM = 0.4;
const TAU = Math.PI * 2;

/** splitmix64 over a pair of 32-bit halves, because JavaScript has no u64.
 *
 * `BigInt` would be the obvious way and is roughly an order of magnitude
 * slower; this runs once per instance per shot, not per frame, but the Atlas
 * has forty thousand instances and "once per instance" is still a number. */
class U64 {
  hi: number;
  lo: number;
  constructor(hi: number, lo: number) {
    this.hi = hi >>> 0;
    this.lo = lo >>> 0;
  }
}

const GAMMA_HI = 0x9e3779b9;
const GAMMA_LO = 0x7f4a7c15;

function add64(a: U64, hi: number, lo: number): void {
  const lo2 = (a.lo >>> 0) + (lo >>> 0);
  a.lo = lo2 >>> 0;
  a.hi = (a.hi + hi + (lo2 > 0xffffffff ? 1 : 0)) >>> 0;
}

function mul64(a: U64, hi: number, lo: number): void {
  // 32x32 -> 64 in 16-bit limbs. The naive `a.lo * lo` loses bits above 2^53.
  const al = a.lo >>> 0;
  const ah = a.hi >>> 0;
  const bl = lo >>> 0;
  const bh = hi >>> 0;
  const a0 = al & 0xffff;
  const a1 = al >>> 16;
  const b0 = bl & 0xffff;
  const b1 = bl >>> 16;

  const p00 = a0 * b0;
  const p01 = a0 * b1;
  const p10 = a1 * b0;
  const p11 = a1 * b1;

  let mid = (p00 >>> 16) + (p01 & 0xffff) + (p10 & 0xffff);
  const resLo = ((mid & 0xffff) << 16) | (p00 & 0xffff);
  let carry = (mid / 0x10000) | 0;
  carry += (p01 >>> 16) + (p10 >>> 16) + p11;
  // The cross terms that only reach the high word.
  const resHi = (carry + Math.imul(al, bh) + Math.imul(ah, bl)) >>> 0;
  a.lo = resLo >>> 0;
  a.hi = resHi >>> 0;
}

function xorShiftRight(a: U64, n: number): void {
  if (n < 32) {
    const lo = ((a.lo >>> n) | (a.hi << (32 - n))) >>> 0;
    const hi = a.hi >>> n;
    a.lo = (a.lo ^ lo) >>> 0;
    a.hi = (a.hi ^ hi) >>> 0;
  } else {
    const lo = a.hi >>> (n - 32);
    a.lo = (a.lo ^ lo) >>> 0;
  }
}

/** One splitmix64 step, returning a double in [0,1) from the top 53 bits —
 * the same construction the Rust side uses, so the two agree exactly. */
export function nextFloat(state: U64): number {
  add64(state, GAMMA_HI, GAMMA_LO);
  const z = new U64(state.hi, state.lo);
  xorShiftRight(z, 30);
  mul64(z, 0xbf58476d, 0x1ce4e5b9);
  xorShiftRight(z, 27);
  mul64(z, 0x94d049bb, 0x133111eb);
  xorShiftRight(z, 31);
  // Top 53 bits: the whole high word plus the top 21 of the low.
  return (z.hi * 0x200000 + (z.lo >>> 11)) / 9007199254740992;
}

/** `SPLITMIX_GAMMA * (index + 1) + editionSeed * 0x2545F4914F6CDD1D`. */
export function placementSeed(index: number, editionSeed: number): U64 {
  const s = new U64(GAMMA_HI, GAMMA_LO);
  mul64(s, 0, (index + 1) >>> 0);
  if (editionSeed) {
    const e = new U64((editionSeed / 0x100000000) >>> 0, editionSeed >>> 0);
    mul64(e, 0x2545f491, 0x4f6cdd1d);
    add64(s, e.hi, e.lo);
  }
  return s;
}

/** Where a placement starts, as an offset from its final position, in LDraw
 * units — Y-down, so the negative Y is upward. */
export function startOffsetLdu(index: number, editionSeed = 0): [number, number, number] {
  const state = placementSeed(index, editionSeed);
  const angle = nextFloat(state) * TAU;
  const radius = SCATTER_RADIUS_LDU * (0.4 + 0.6 * nextFloat(state));
  return [radius * Math.cos(angle), -FLOAT_HEIGHT_LDU, radius * Math.sin(angle)];
}

/** The same offset in the viewer's frame: millimetres, +Y up.
 *
 * The Y negation is the mirror `spex_mesh::bundle::to_output_position`
 * applies to every vertex in the library, and it is applied here for the
 * same reason — the two frames differ by exactly this. */
export function startOffsetMm(index: number, editionSeed = 0): [number, number, number] {
  const [x, y, z] = startOffsetLdu(index, editionSeed);
  return [x * LDU_MM, -y * LDU_MM, z * LDU_MM];
}

/** A placement's own progress, given the shot's. See the Rust twin. */
export function staggeredProgress(t01: number, order: number, count: number, stagger: number): number {
  const t = t01 < 0 ? 0 : t01 > 1 ? 1 : t01;
  if (count <= 1 || stagger <= 0) return t;
  const s = stagger > 1 ? 1 : stagger;
  const span = 1 - s;
  const start = (s * Math.min(order, count - 1)) / (count - 1);
  if (span <= Number.EPSILON) return t >= start ? 1 : 0;
  const u = (t - start) / span;
  return u < 0 ? 0 : u > 1 ? 1 : u;
}

/** Anything that can be told where an instance is. `InstanceWriter` from
 * M55 satisfies this; so does a recording object in a test harness, which is
 * how this module gets verified without a scene. */
export interface TransformTarget {
  setTransform(id: string, position: PositionLike, quaternion: QuaternionLike, scale: number): void;
}
export interface PositionLike {
  set(x: number, y: number, z: number): unknown;
  x: number;
  y: number;
  z: number;
}
export interface QuaternionLike {
  set(x: number, y: number, z: number, w: number): unknown;
}

export interface AssemblySpec {
  /** Instance ids, in bundle order, that this assembly moves. */
  ids: readonly string[];
  /** Each instance's final position in millimetres, +Y up — where the bundle
   * already placed it. */
  finals: Float32Array | readonly number[];
  /** Rank per instance, deciding the stagger order. The real `0 STEP` build
   * step where the scene has one; the index otherwise. */
  order?: readonly number[];
  editionSeed?: number;
  /** Fraction of the shot spent handing over from the first part to the last. */
  stagger?: number;
  easing?: EasingFn;
}

/** How much of a shot the first part spends landing before the last begins.
 *
 * 0.55 rather than a round number: the assembly is four bars, and the
 * screenplay's direction is that each landing is its own accent — too much
 * stagger and the last brick arrives after the music has moved on, too
 * little and nine accents become one. */
export const DEFAULT_STAGGER = 0.55;

/** The scattered-start assembly, evaluated.
 *
 * Holds the start offsets it generated so the per-frame path is a lerp and
 * nothing else: seeding once per instance per *shot* rather than per frame is
 * the difference between this costing nothing and it costing 40 000 splitmix
 * chains a frame at Atlas scale.
 */
export class AssemblyChoreography {
  readonly count: number;
  private readonly ids: readonly string[];
  private readonly finals: Float32Array;
  private readonly starts: Float32Array;
  private readonly order: readonly number[];
  private readonly stagger: number;
  private readonly ease: EasingFn;
  private readonly ranks: number[];
  /** How many *distinct* build steps the scene has — the span the stagger
   * divides, which is not the instance count when steps repeat. */
  private readonly rankCount: number;

  constructor(spec: AssemblySpec) {
    this.ids = spec.ids;
    this.count = spec.ids.length;
    this.finals = spec.finals instanceof Float32Array ? spec.finals : Float32Array.from(spec.finals);
    this.stagger = spec.stagger ?? DEFAULT_STAGGER;
    this.ease = spec.easing ?? cubicInOut;
    this.order = spec.order ?? spec.ids.map((_, i) => i);

    // Build steps are not dense: a scene may have steps 0, 0, 1, 3. Ranking
    // them turns "which step" into "how far through the build", which is what
    // the stagger actually wants — otherwise a scene whose steps happen to be
    // sparse staggers over a shorter span than one whose steps are dense, for
    // no reason a viewer could perceive.
    const distinct = Array.from(new Set(this.order)).sort((a, b) => a - b);
    const rankOf = new Map(distinct.map((v, i) => [v, i]));
    this.ranks = this.order.map((v) => rankOf.get(v)!);
    this.rankCount = distinct.length;

    this.starts = new Float32Array(this.count * 3);
    for (let i = 0; i < this.count; i++) {
      const [dx, dy, dz] = startOffsetMm(i, spec.editionSeed ?? 0);
      this.starts[i * 3] = this.finals[i * 3] + dx;
      this.starts[i * 3 + 1] = this.finals[i * 3 + 1] + dy;
      this.starts[i * 3 + 2] = this.finals[i * 3 + 2] + dz;
    }
  }

  /** Where instance `i` is at shot progress `t01`, written into `out`. */
  positionAt(i: number, t01: number, out: [number, number, number]): [number, number, number] {
    const u = this.ease(staggeredProgress(t01, this.ranks[i], this.rankCount, this.stagger));
    const o = i * 3;
    out[0] = this.starts[o] + (this.finals[o] - this.starts[o]) * u;
    out[1] = this.starts[o + 1] + (this.finals[o + 1] - this.starts[o + 1]) * u;
    out[2] = this.starts[o + 2] + (this.finals[o + 2] - this.starts[o + 2]) * u;
    return out;
  }

  /** Writes every instance's position for this frame.
   *
   * `position` and `quaternion` are the caller's scratch objects, reused —
   * see `InstanceWriter`, which makes the same bargain for the same reason. */
  apply(t01: number, target: TransformTarget, position: PositionLike, quaternion: QuaternionLike): void {
    quaternion.set(0, 0, 0, 1);
    for (let i = 0; i < this.count; i++) {
      const u = this.ease(staggeredProgress(t01, this.ranks[i], this.rankCount, this.stagger));
      const o = i * 3;
      position.set(
        this.starts[o] + (this.finals[o] - this.starts[o]) * u,
        this.starts[o + 1] + (this.finals[o + 1] - this.starts[o + 1]) * u,
        this.starts[o + 2] + (this.finals[o + 2] - this.starts[o + 2]) * u,
      );
      target.setTransform(this.ids[i], position, quaternion, 1);
    }
  }
}

/** Reads an assembly out of a shot's `seed` cue payload.
 *
 * The document says *what* to generate and this decides *how*, which is the
 * split that keeps five thousand transform tracks out of the resolved file.
 * An unknown generator name returns null rather than throwing: a document
 * from a later version of the piece should degrade to "that shot does not
 * animate", not to a blank page. */
export function assemblyFromCue(
  payload: Record<string, unknown> | undefined,
  spec: Omit<AssemblySpec, 'stagger' | 'easing' | 'editionSeed'>,
  editionSeed: number,
): AssemblyChoreography | null {
  if (!payload || payload.generator !== 'assembly') return null;
  const easingName = typeof payload.easing === 'string' ? payload.easing : 'cubicInOut';
  return new AssemblyChoreography({
    ...spec,
    editionSeed,
    stagger: typeof payload.stagger === 'number' ? payload.stagger : DEFAULT_STAGGER,
    easing: easingByName(easingName),
  });
}

File: spex/crates/spex-show/src/choreography.rs
//! M64 — the generated choreography, and the one PRNG both sides can agree on.
//!
//! `spex brick-assembly` has been settling nine bricks out of a scattered
//! start since long before the Fugen Engine existed, and A1-S04 is that shot.
//! The milestone's job is to stop *baking* it into frames and evaluate it at
//! runtime instead — which means the browser has to produce the identical
//! starting layout the Rust demo does.
//!
//! # Why `rand::StdRng` had to go
//!
//! `brick.rs` seeded `StdRng::seed_from_u64` per placement. `StdRng` is
//! ChaCha12: a cryptographic stream, reimplementable in TypeScript only at
//! real cost, and — this is the part that settles it — **`rand` explicitly
//! does not promise its output is stable across versions.** A piece whose
//! visual choreography changes when a transitive dependency bumps a minor
//! version is not reproducible, and "the 2027 edition looked different after
//! `cargo update`" is not a sentence anyone wants to write.
//!
//! So the generator moved to **splitmix64** — twelve lines, a fixed
//! specification, bit-identical in any language with a u64. The old code's
//! comment already said "a splitmix-style constant"; it used that constant to
//! seed ChaCha. Now it is splitmix all the way through.
//!
//! **This changes the existing demo's scatter.** The parts start from
//! different directions than they used to. That is cosmetic — the layout was
//! always arbitrary-but-deterministic, and it is still arbitrary-but-
//! deterministic — and it buys the property the whole milestone rests on:
//! baked and runtime agree *by construction* rather than by coincidence.
//!
//! # Coordinates
//!
//! Everything here is in **LDraw units, Y-down**, because that is the frame
//! the constants were measured in and the frame `brick.rs` works in. The
//! viewer converts once, at the same boundary everything else does.

/// How far "up" (LDraw −Y) each part starts before settling. Ported, not
/// re-chosen: the existing demo's look depends on it.
pub const FLOAT_HEIGHT_LDU: f64 = 420.0;
/// Deterministic sideways scatter, so parts converge from different
/// directions rather than all dropping in a vertical line.
pub const SCATTER_RADIUS_LDU: f64 = 260.0;
/// The golden-ratio odd constant splitmix64 is defined with.
pub const SPLITMIX_GAMMA: u64 = 0x9E37_79B9_7F4A_7C15;

/// One step of splitmix64. Advances `state` and returns the mixed output.
#[inline]
pub fn splitmix64(state: &mut u64) -> u64 {
    *state = state.wrapping_add(SPLITMIX_GAMMA);
    let mut z = *state;
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    z ^ (z >> 31)
}

/// A double in [0,1), from the top 53 bits — the same construction
/// `Math.random()` is specified to produce, so the JavaScript port is exact
/// rather than approximately exact.
#[inline]
pub fn next_f64(state: &mut u64) -> f64 {
    (splitmix64(state) >> 11) as f64 / (1u64 << 53) as f64
}

/// The seed for one placement. Index-derived, so a scene's layout does not
/// change when a *different* scene is built, and edition-seeded, so the
/// endless edition's cycle advance actually reaches the choreography.
#[inline]
pub fn placement_seed(index: usize, edition_seed: u64) -> u64 {
    SPLITMIX_GAMMA
        .wrapping_mul(index as u64 + 1)
        .wrapping_add(edition_seed.wrapping_mul(0x2545_F491_4F6C_DD1D))
}

/// Where a placement starts, as an offset from its own final position, in
/// LDraw units (Y-down: the negative Y is *upward*).
pub fn start_offset_ldu(index: usize, edition_seed: u64) -> [f64; 3] {
    let mut state = placement_seed(index, edition_seed);
    let angle = next_f64(&mut state) * std::f64::consts::TAU;
    let radius = SCATTER_RADIUS_LDU * (0.4 + 0.6 * next_f64(&mut state));
    [radius * angle.cos(), -FLOAT_HEIGHT_LDU, radius * angle.sin()]
}

/// The curve the assembly settles on, and the reference implementation the
/// viewer's `easing.ts` is a port of.
///
/// It lived in `brick.rs` as `ease_in_out_cubic`. It is here now for the same
/// reason the scatter is: two implementations of one curve agree to about
/// three decimals, and this milestone's whole claim is that the baked demo
/// and the runtime shot produce the *same* positions.
#[inline]
pub fn cubic_in_out(t: f64) -> f64 {
    let t = t.clamp(0.0, 1.0);
    if t < 0.5 {
        4.0 * t * t * t
    } else {
        1.0 - (-2.0 * t + 2.0).powi(3) / 2.0
    }
}

/// A placement's own progress through an assembly, given the shot's progress.
///
/// `stagger` is the fraction of the shot spent handing over from the first
/// part to the last: at 0 everything lands together, at 1 the last part does
/// not begin until the first has finished. `order` is the placement's rank —
/// its index, or its real build step where the scene has `0 STEP` lines,
/// which is the whole reason those were parsed.
///
/// Returns raw 0..1 progress; the caller eases it. Keeping the easing out
/// means the same stagger works for a curve authored per shot.
pub fn staggered_progress(t01: f64, order: usize, count: usize, stagger: f64) -> f64 {
    if count <= 1 || stagger <= 0.0 {
        return t01.clamp(0.0, 1.0);
    }
    let s = stagger.clamp(0.0, 1.0);
    let span = 1.0 - s;
    // Each placement's window is `span` long and starts `s * rank/(n-1)` in.
    let start = s * (order.min(count - 1) as f64) / ((count - 1) as f64);
    if span <= f64::EPSILON {
        return if t01 >= start { 1.0 } else { 0.0 };
    }
    ((t01 - start) / span).clamp(0.0, 1.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// splitmix64's first outputs from state 0 are published reference values.
    /// If this drifts, every edition of the piece drifts with it.
    #[test]
    fn splitmix64_matches_its_published_reference_stream() {
        let mut s = 0u64;
        assert_eq!(splitmix64(&mut s), 0xE220_A839_7B1D_CDAF);
        assert_eq!(splitmix64(&mut s), 0x6E78_9E6A_A1B9_65F4);
        assert_eq!(splitmix64(&mut s), 0x06C4_5D18_8009_454F);
    }

    #[test]
    fn a_double_from_splitmix_is_in_range_and_reproducible() {
        let mut a = 12345u64;
        let mut b = 12345u64;
        for _ in 0..1000 {
            let x = next_f64(&mut a);
            assert!((0.0..1.0).contains(&x), "{x} outside [0,1)");
            assert_eq!(x, next_f64(&mut b));
        }
    }

    /// The scatter is a *ring*, not a disc, and both bounds matter: the inner
    /// one keeps a part from starting on top of its own destination, the
    /// outer one keeps it in frame.
    #[test]
    fn every_start_offset_lands_in_the_authored_ring_and_at_the_authored_height() {
        for i in 0..500 {
            let [x, y, z] = start_offset_ldu(i, 0);
            let r = (x * x + z * z).sqrt();
            assert!(
                r >= SCATTER_RADIUS_LDU * 0.4 - 1e-9 && r <= SCATTER_RADIUS_LDU + 1e-9,
                "placement {i}: radius {r}"
            );
            assert_eq!(y, -FLOAT_HEIGHT_LDU);
        }
    }

    /// The edition seed has to actually reach the choreography, or an
    /// "edition" is the same film with a different number printed on it.
    #[test]
    fn a_different_edition_seed_produces_a_different_layout() {
        let a: Vec<[f64; 3]> = (0..9).map(|i| start_offset_ldu(i, 0)).collect();
        let b: Vec<[f64; 3]> = (0..9).map(|i| start_offset_ldu(i, 263865)).collect();
        assert_ne!(a, b);
        // ...but still deterministic within one edition.
        let a2: Vec<[f64; 3]> = (0..9).map(|i| start_offset_ldu(i, 0)).collect();
        assert_eq!(a, a2);
    }

    #[test]
    fn stagger_hands_over_from_the_first_placement_to_the_last() {
        // With no stagger everything moves together.
        assert_eq!(staggered_progress(0.5, 0, 9, 0.0), 0.5);
        assert_eq!(staggered_progress(0.5, 8, 9, 0.0), 0.5);

        // With stagger the last placement has not started when the first is
        // already half done.
        let first = staggered_progress(0.3, 0, 9, 0.6);
        let last = staggered_progress(0.3, 8, 9, 0.6);
        assert!(first > last, "{first} should lead {last}");
        assert_eq!(last, 0.0, "the last placement waits");

        // Everything is finished at the end, and nothing before the start.
        for order in 0..9 {
            assert_eq!(staggered_progress(0.0, order, 9, 0.6), if order == 0 { 0.0 } else { 0.0 });
            assert_eq!(staggered_progress(1.0, order, 9, 0.6), 1.0);
        }
    }

    /// The fixture both languages are pinned to. If this test fails, the Rust
    /// generator changed; if the browser check fails against the same file,
    /// the TypeScript one did. Either way the two have stopped agreeing, and
    /// that is the thing M64 exists to prevent.
    #[test]
    fn the_shared_fixture_still_describes_this_generator() {
        let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../docs/fugen/fixtures/assembly-scatter.json");
        let text = std::fs::read_to_string(&path)
            .unwrap_or_else(|e| panic!("reading {}: {e}", path.display()));
        let fixture: serde_json::Value = serde_json::from_str(&text).unwrap();

        for case in fixture["cases"].as_array().unwrap() {
            let seed = case["editionSeed"].as_u64().unwrap();
            let expected = case["offsetsLdu"].as_array().unwrap();
            for (i, want) in expected.iter().enumerate() {
                let got = start_offset_ldu(i, seed);
                for k in 0..3 {
                    let w = want[k].as_f64().unwrap();
                    assert!(
                        (got[k] - w).abs() < 1e-9,
                        "seed {seed}, placement {i}, component {k}: {} vs fixture {w}",
                        got[k]
                    );
                }
            }
        }
    }
}
